### PR TITLE
Adding Achievements Flat C

### PR DESCRIPTION
**Pull request by rakhad**

### CODE REVIEW

File: Include/xsapi-c/achievements_c.h [hunk 1]
```diff
@@ -11,234 +11,247 @@
...
-struct XSAPI_ACHIEVEMENT_TITLE_ASSOCIATION_IMPL;
-struct XSAPI_ACHIEVEMENT_REQUIREMENT_IMPL;
-struct XSAPI_ACHIEVEMENT_PROGRESSION_IMPL;
-struct XSAPI_ACHIEVEMENT_MEDIA_ASSET_IMPL;
-struct XSAPI_ACHIEVEMENT_REWARD_IMPL;
-struct XSAPI_ACHIEVEMENT_IMPL;
-struct XSAPI_ACHIEVEMENTS_RESULT_IMPL;
-
-typedef enum XSAPI_ACHIEVEMENT_TYPE
+struct XBL_ACHIEVEMENT_TITLE_ASSOCIATION_IMPL;
```
Comment (jplafonta) [inline Include/xsapi-c/achievements_c.h]: You will need to go back and change all these names align with the gamecore standards.  You can do it now or in a future check-in.

File: Include/xsapi-c/achievements_c.h [hunk 2]
```diff
@@ -11,234 +11,247 @@
...
+struct XBL_ACHIEVEMENTS_RESULT_IMPL;
+
+typedef enum XBL_ACHIEVEMENT_TYPE
 {
-    XSAPI_ACHIEVEMENT_TYPE_UNKNOWN,
-    XSAPI_ACHIEVEMENT_TYPE_ALL,
-    XSAPI_ACHIEVEMENT_TYPE_PERSISTENT,
-    XSAPI_ACHIEVEMENT_TYPE_CHALLENGE
-} XSAPI_ACHIEVEMENT_TYPE;
+    XBL_ACHIEVEMENT_TYPE_UNKNOWN,
```
Comment (jplafonta) [inline Include/xsapi-c/achievements_c.h]: also add doc comments similar to the C++ headers

File: Include/xsapi-c/achievements_c.h [hunk 3]
```diff
@@ -11,234 +11,247 @@
...
 
-    XSAPI_ACHIEVEMENT_TITLE_ASSOCIATION_IMPL* pImpl;
-} XSAPI_ACHIEVEMENT_TITLE_ASSOCIATION;
+    XBL_ACHIEVEMENT_TITLE_ASSOCIATION_IMPL* pImpl;
+} XBL_ACHIEVEMENT_TITLE_ASSOCIATION;
 
-typedef struct XSAPI_ACHIEVEMENT_REQUIREMENT
+typedef struct XBL_ACHIEVEMENT_REQUIREMENT
 {
     PCSTR id;
```
Comment (jplafonta) [inline Include/xsapi-c/achievements_c.h]: If possible make all struct fields fixed sizes so that copying becomes easier. ie char id[MAX_SIZE]

**Comment by jplafonta** [comment]
> #pragma once

#include this file in services_c.h

---
Refers to: Include/xsapi-c/achievements_c.h:4 in a330c11. [](commit_id = a330c1138b2b77ceecb64d2c45249f1d7e4f657d, deletion_comment = False)

File: Source/Services/Achievements/C/achievements_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+XBL_ACHIEVEMENT* CreateAchievementFromCpp(
+    _In_ std::shared_ptr<achievement_internal> cppAchievement
+)
+{
+    auto singleton = get_singleton_for_achievements();
+    std::lock_guard<std::recursive_mutex> lock(singleton->m_achievementsState->m_lock);
+
+    auto achievement = new XBL_ACHIEVEMENT();
```
Comment (jplafonta) [inline Source/Services/Achievements/C/achievements_helper.cpp]: Where does this get cleaned up?  Also you will need to use memhooks here and everywhere else you are using "new"

File: Include/xsapi-c/achievements_c.h (hunk 3 above)
Comment (rakhad) [reply Include/xsapi-c/achievements_c.h]: MAX_SIZE for the field?

---
In reply to: [181464782](https://github.com/Microsoft/xbox-live-api/pull/340#discussion_r181464782) [](ancestors = 181464782)

**Pull request merged**